### PR TITLE
[main][feat] xALPACACreditor value setter

### PR DESCRIPTION
**Pull request by wow-alpaca**
## Description
- Add value setter in xALPACACreditor contract

## Why?
- Automated xALPACA value setter

## ChangeLogs:
- `xALPACACreditor.sol`
- `xALPACACreditorLike.sol`
- `xALPACACreditor.t.sol`

### CODE REVIEW

File: solidity/contracts/8.13/xALPACACreditor.sol [hunk 1]
```diff
@@ -49,9 +52,18 @@ contract xALPACACreditor is OwnableUpgradeable, ICreditor {
     return (xALPACA.balanceOf(_user) * valuePerxALPACA) / 1e18;
   }
 
+  /// @notice set the value setter
+  function setValueSetter(address _valueSetter) external onlyOwner {
```
Comment (alpanaca) [inline solidity/contracts/8.13/xALPACACreditor.sol:56]:
```suggestion
  function setValueSetter(address _newValueSetter) external onlyOwner {
```

File: solidity/contracts/8.13/xALPACACreditor.sol [hunk 2]
```diff
@@ -22,13 +22,16 @@ import { ICreditor } from "./interfaces/ICreditor.sol";
 contract xALPACACreditor is OwnableUpgradeable, ICreditor {
   // --- Events ---
   event LogSetValuePerxALPACA(address indexed _caller, uint256 _oldValuePerxALPACA, uint256 _newValuePerxALPACA);
+  event LogSetValueSetter(address indexed _caller, address indexed _valueSetter);
 
   // --- Errors ---
   error xALPACACreditor_ValueTooHigh();
+  error xALPACACreditor_SenderIsNotValueSetter();
```
Comment (alpanaca) [inline solidity/contracts/8.13/xALPACACreditor.sol:29]:
```suggestion
  error xALPACACreditor_Unauthorize();
```

File: solidity/contracts/8.13/xALPACACreditor.sol [hunk 3]
```diff
@@ -49,9 +52,18 @@ contract xALPACACreditor is OwnableUpgradeable, ICreditor {
...
+    emit LogSetValueSetter(msg.sender, _valueSetter);
+  }
+
   /// @notice Set the value per xALPACA
   /// @param _newValuePerxALPACA new value to be set.
-  function setValuePerxALPACA(uint256 _newValuePerxALPACA) external onlyOwner {
+  function setValuePerxALPACA(uint256 _newValuePerxALPACA) external {
+    if (msg.sender != valueSetter) {
+      revert xALPACACreditor_SenderIsNotValueSetter();
+    }
```
Comment (alpanaca) [inline solidity/contracts/8.13/xALPACACreditor.sol:66]:
```suggestion
    if (msg.sender != valueSetter) {
      revert xALPACACreditor_Unauthorize();
    }
```

File: solidity/tests/interfaces/xALPACACreditorLike.sol [hunk 4]
```diff
@@ -17,4 +17,6 @@ interface xALPACACreditorLike {
   function getUserCredit(address _user) external view returns (uint256);
 
   function setValuePerxALPACA(uint256 _newValuePerxALPACA) external;
+
+  function setValueSetter(address _valueSetter) external;
```
Comment (alpanaca) [inline solidity/tests/interfaces/xALPACACreditorLike.sol:21]:
```suggestion
  function setValueSetter(address _newValueSetter) external;
```

File: solidity/tests/protocol/private-automated-vault/xALPACACreditor.t.sol [hunk 5]
```diff
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: MIT
 pragma solidity >=0.8.4 <0.9.0;
 
-import { BaseTest, xALPACACreditorLike } from "../../base/BaseTest.sol";
+import { BaseTest, xALPACACreditorLike, console } from "../../base/BaseTest.sol";
```
Comment (alpanaca) [inline solidity/tests/protocol/private-automated-vault/xALPACACreditor.t.sol:4]: remove console

File: solidity/tests/protocol/private-automated-vault/xALPACACreditor.t.sol [hunk 6]
```diff
@@ -24,31 +24,41 @@ contract xAlpacaCreditor_Test is BaseTest {
...
   }
 
   function testCanSetValuePerXalpacaLessThanThreshold(uint256 _value) external {
     vm.assume(_value < 1000 ether);
+    vm.prank(ALICE);
     _creditor.setValuePerxALPACA(_value);
   }
+
+  function testCannotSetValuePerXalpacaIfNotValueSetter() external {
+    vm.expectRevert(abi.encodeWithSignature("xALPACACreditor_SenderIsNotValueSetter()"));
```
Comment (alpanaca) [inline solidity/tests/protocol/private-automated-vault/xALPACACreditor.t.sol:59]:
```suggestion
    vm.expectRevert(abi.encodeWithSignature("xALPACACreditor_Unauthorize()"));
```

File: solidity/contracts/8.13/xALPACAPriceSetter.sol [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+import { IxALPACACreditor } from "./interfaces/IxALPACACreditor.sol";
+import { ITWAPOracle } from "./interfaces/ITWAPOracle.sol";
+
+/// @title xALPACAPriceSetter - Being setter of xALPACACreditor
+contract xALPACAPriceSetter is OwnableUpgradeable {
+  // --- Events ---
+  event LogSetValueFromTWAP(address indexed _caller, uint256 _xALPACAValue);
+
+  // --- Errors ---
```
Comment (alpanaca) [inline solidity/contracts/8.13/xALPACAPriceSetter.sol:26]: Just remove this if we don't have the error.

File: solidity/contracts/8.13/xALPACAPriceSetter.sol [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  function initialize(IxALPACACreditor _xALPACACreditor, ITWAPOracle _TWAPOracle, address _alpaca) external initializer {
+
+    OwnableUpgradeable.__Ownable_init();
+    xALPACACreditor = IxALPACACreditor(_xALPACACreditor);
+    TWAPOracle = _TWAPOracle;
+    alpaca = _alpaca;
+  }
+
+  /// @notice Set ALPACA Value (TWAP) as xALPACA Value
+  function setValueFromTWAP() external onlyOwner {
```
Comment (alpanaca) [inline solidity/contracts/8.13/xALPACAPriceSetter.sol:46]: Have a whitelist on this. I don't think the deployer will be the one who call this.

**Review by abc-alpaca — APPROVED** [review]

**Review by alpanaca — APPROVED** [review]

**Pull request merged**